### PR TITLE
Web directory structure adr proposal

### PR DESCRIPTION
**Pull request by CDimonaco**
This is a proposal for refactoring the current directory structure of the `web` repository.

This is a general guideline, the moving and refactoring of each file/module is delegated to the implementation, this should be sufficient for guiding the developer during the process

### CODE REVIEW

**Comment by dottorblaster** [comment]
I'M LOVIN IT

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+## Context
+
+The web dashboard repository, is a `phoenix` application, right now the directory structure and the overall `phoenix` framework convention regarding modules and code organization are not respected.
+With the increasing complexity of the project and the codebase, we need to switch to the `phoenix` framework standards, to be compatible with code generators and framework conventions in general.
+
+The current directory structure is too complex in a lot of aspects, given the fact that it's really difficult to follow where the moving parts are and deciding where to put what is becoming more difficult with the increasing codebase complexity.
+
+The web project is also a `commanded` application, `commanded` has examples and guidelines on how to structure the projects, the directories and the modules, this documentation with the official  `phoenix` guide will be the base principles for this refactoring.
+As software engineers, we propose a new structure/architecture to increase developers' productivity and confidence and obtain adherence to framework/language standards.
```
Comment (dottorblaster) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:17]:
```suggestion
As software engineers, we propose a new structure/architecture to increase developers' productivity, confidence, and obtain adherence to framework/language standards.

As a side effect, we will align Wanda and Web to the same standard, decreasing onboarding length.
```

**Review by dottorblaster — COMMENTED** [review]
Just this

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 2]
```diff
@@ -0,0 +1,79 @@
+# 10. Web dashboard directory structure and contexts
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:1]: Maybe change to `10. Web repository directory...`

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+
+Date: 2023-08-24
+
+## Status
+
+Proposal
+
+## Context
+
+The web dashboard repository, is a `phoenix` application, right now the directory structure and the overall `phoenix` framework convention regarding modules and code organization are not respected.
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:11]: `The web dashboard repository is a phoenix application. Right now, the directory`. A `.` looks better there.

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 4]
```diff
@@ -0,0 +1,79 @@
...
+With the increasing complexity of the project and the codebase, we need to switch to the `phoenix` framework standards, to be compatible with code generators and framework conventions in general.
+
+The current directory structure is too complex in a lot of aspects, given the fact that it's really difficult to follow where the moving parts are and deciding where to put what is becoming more difficult with the increasing codebase complexity.
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:14]: `...and deciding where to put them, which is becoming even more difficult...`

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+
+The web project is also a `commanded` application, `commanded` has examples and guidelines on how to structure the projects, the directories and the modules, this documentation with the official  `phoenix` guide will be the base principles for this refactoring.
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:16]: `The web repository is also a commended application. commanded has examples...`. Better a dot after the 1st phrase.

`...projects, the directories and the modules. This documentation`. Better dot here as well

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+- https://github.com/slashdotdash/segment-challenge/tree/master/lib/segment_challenge/challenges
+- https://github.com/slashdotdash/segment-challenge/tree/master/lib/segment_challenge/athletes
+- https://github.com/slashdotdash/segment-challenge/tree/master/lib/segment_challenge/infrastructure
+- https://github.com/slashdotdash/segment-challenge/tree/master/lib/segment_challenge/stages
+
+## Decision
+
+We propose a directory structure with these characteristics
+
+- One context for each "domain entity" of our application (cluster, sap_systems, hosts, tags etc..)
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:33]: `cluster` is singular and the other plural. Let's put them in the same way hehe

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+- Everything that is tailored to the context in the same directory, (policies, "services", etc..)
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:34]: Why `services` with quotes?

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+- "integration"/"infrastructure" context, containing all the interaction outside the domain like rabbitmq, checks engine etc.. (what we have now in application directory) and commanded cross-aggregate interactions
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:35]: `checks engines, etc`

File: adr/0010-web-dashboard-directory-structure-and-contexts.md [hunk 9]
```diff
@@ -0,0 +1,80 @@
...
+            grafana
+            prometheus
+            telemetry
+            auth
+```
+
+
+The `trento_web` directory should contain all the files and contexts dictated by the `phoenix` guide, including mailer and mail templates. 
+
+The module names should be refactored to reflect the directory structure, this means that we will need an ecto migration, to change module names in the commanded persistence, to maintain compatibility between old/new installations
```
Comment (arbulu89) [inline adr/0010-web-dashboard-directory-structure-and-contexts.md:74]: `...the directory structure. This means that...`

**Review by arbulu89 — COMMENTED** [review]
The content looks great.
Some questions:
- Where do we put generic files like `integration/usecases/roll_up`? Or `tags`. We have some code associated to things that are neither commanded or pheonix contexts.
- In the given examples, the `queries` folder contains individual files for individual queries. Are we going to follow that rule?
- There are many other folders like `payloads`, `value_objects`, `enums` which I wouldn't know where they should go

PD: Just some improvements in the text.

**Comment by CDimonaco** [comment]
> The content looks great. Some questions:
> 
>     * Where do we put generic files like `integration/usecases/roll_up`? Or `tags`. We have some code associated to things that are neither commanded or pheonix contexts.

For the rollup, we put that in integration/commanded, is related to commanded but not tailored directly to our aggregates/domain.

Tags are a separate context, in this case they are the "canonical" phoenix context, because we have pure crud and logic, controller, model, queries etc..
> 
>     * In the given examples, the `queries` folder contains individual files for individual queries. Are we going to follow that rule?

I think it's not a strict rule, I saw both one file queries, or one file with each file, this is just a guideline, we don't have right know too much custom queries to even justify the queries module tbh. Only the rollup has some custom queries but it goes in the integration/commanded rollup with other rollup related things.
> 
>     * There are many other folders like `payloads`, `value_objects`, `enums` which I wouldn't know where they should go
> 
They should be located near the context they belong, for example `AscsErsClusterDetails` in the cluster context

The general rule is to have all in the appropriate context, for some cross context stuff we can discuss case by case, infact it's a very discussed topic even in the phoenix community itself, but as I looked into the codebase I think that is pretty much tailored to some context, we have only in application some cross concern that could be easily translated into the infrastructure/integration module.


This is just a guidance, if we have some edge case or some files/modules that should be particular discussed we can discuss that on the implementation, maybe they need to be refactored completely or we can find some temporary solution for very peculiar stuff.

**Review by arbulu89 — APPROVED** [review]
Thank you @CDimonaco 
I like the idea of moving to a "phoenix" standards directory structure.
We will see what we do with the "edge cases" when we start working on them

File: adr/0010-web-dashboard-directory-structure-and-contexts.md (hunk 7 above)
Comment (CDimonaco) [reply adr/0010-web-dashboard-directory-structure-and-contexts.md:34]: tbh idk lol fixed

**Comment by CDimonaco** [comment]
> Thank you @CDimonaco I like the idea of moving to a "phoenix" standards directory structure. We will see what we do with the "edge cases" when we start working on them

Exactly, thanks for the feedback, ofc when we will work on the implementation, we will create separate pr for each context or at least for each "part" we want to refactor, so we can address all the particular concerns one-by-one.

**Review by dottorblaster — APPROVED** [review]
:heart_eyes:

**Review by nelsonkopliku — APPROVED** [review]
This clears all the doubts I had about the direction of this refactor. Thanks a lot!
I actually like the outline that it gets.

I also totally agree in handling specific edge cases when they rise.

Well done!

**Review by EMaksy — APPROVED** [review]
Hey Thanks for that :D LGTM :+1:

**Pull request merged**